### ISSUE TITLE
BUG：Flutter will call addActivityResultListener twice in flutterEmbedding v2

### ISSUE THREAD

**Issue by v7lin** [flutter/flutter#49105]
[fake_tencent/issues/24](https://github.com/v7lin/fake_tencent/issues/24)

v2 plugins
```java
/**
 * TencentKitPlugin
 */
public class TencentKitPlugin implements FlutterPlugin, ActivityAware {
    // This static function is optional and equivalent to onAttachedToEngine. It supports the old
    // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
    // plugin registration via this function while apps migrate to use the new Android APIs
    // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
    //
    // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
    // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
    // depending on the user's project. onAttachedToEngine or registerWith must both be defined
    // in the same class.
    public static void registerWith(Registrar registrar) {
        TencentKit tencentKit = new TencentKit(registrar.context(), registrar.activity());
        registrar.addActivityResultListener(tencentKit);
        tencentKit.startListening(registrar.messenger());
    }

    // --- FlutterPlugin

    private final TencentKit tencentKit;

    private ActivityPluginBinding pluginBinding;

    public TencentKitPlugin() {
        tencentKit = new TencentKit();
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        tencentKit.setApplicationContext(flutterPluginBinding.getApplicationContext());
        tencentKit.setActivity(null);
        tencentKit.startListening(flutterPluginBinding.getBinaryMessenger());
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        tencentKit.setApplicationContext(null);
        tencentKit.setActivity(null);
        tencentKit.stopListening();
    }

    // --- ActivityAware

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        tencentKit.setActivity(binding.getActivity());
        pluginBinding = binding;
        pluginBinding.addActivityResultListener(tencentKit);
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        tencentKit.setActivity(null);
        pluginBinding.removeActivityResultListener(tencentKit);
        pluginBinding = null;
    }
}
```

or

v1 plugins
```java
public class TencentKitPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener {
    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "v7lin.github.io/tencent_kit");
        TencentKitPlugin plugin = new TencentKitPlugin(registrar, channel);
        registrar.addActivityResultListener(plugin);
        channel.setMethodCallHandler(plugin);
    }
    ...
}
```

**Comment by v7lin** [comment]
FlutterActivityAndFragmentDelegate.java

```java
  void onAttach(@NonNull Context context) {
    ensureAlive();

    // When "retain instance" is true, the FlutterEngine will survive configuration
    // changes. Therefore, we create a new one only if one does not already exist.
    if (flutterEngine == null) {
      setupFlutterEngine();
    }

    // Regardless of whether or not a FlutterEngine already existed, the PlatformPlugin
    // is bound to a specific Activity. Therefore, it needs to be created and configured
    // every time this Fragment attaches to a new Activity.
    // TODO(mattcarroll): the PlatformPlugin needs to be reimagined because it implicitly takes
    //                    control of the entire window. This is unacceptable for non-fullscreen
    //                    use-cases.
    platformPlugin = host.providePlatformPlugin(host.getActivity(), flutterEngine);

    if (host.shouldAttachEngineToActivity()) {
      // Notify any plugins that are currently attached to our FlutterEngine that they
      // are now attached to an Activity.
      //
      // Passing this Fragment's Lifecycle should be sufficient because as long as this Fragment
      // is attached to its Activity, the lifecycles should be in sync. Once this Fragment is
      // detached from its Activity, that Activity will be detached from the FlutterEngine, too,
      // which means there shouldn't be any possibility for the Fragment Lifecycle to get out of
      // sync with the Activity. We use the Fragment's Lifecycle because it is possible that the
      // attached Activity is not a LifecycleOwner.
      Log.d(TAG, "Attaching FlutterEngine to the Activity that owns this Fragment.");
      flutterEngine.getActivityControlSurface().attachToActivity(
          host.getActivity(),
          host.getLifecycle()
      );
    }

    host.configureFlutterEngine(flutterEngine); 
  }
```

both 
```java
      flutterEngine.getActivityControlSurface().attachToActivity(
          host.getActivity(),
          host.getLifecycle()
      );
```
and
```
host.configureFlutterEngine(flutterEngine); 
```
will call FlutterEnginePluginRegistry#attachToActivity
```
 activityAware.onAttachedToActivity(activityPluginBinding);
```

**Comment by v7lin** [comment]
Any news update?

**Comment by matthew-carroll** [comment]
Thanks for filing this issue. I took a glance at the source code and it does seem like double invocation might be happening. I'll file this ticket in my backlog to investigate.

**Comment by mikemaksymowych** [comment]
I'm seeing this issue too, which I believe is causing my flutter_local_notifications plugin to fire my notification handler callback twice.

**Comment by matthew-carroll** [comment]
@v7lin on a 2nd look at this, I'm not sure I see the problem.

You said

> both
    flutterEngine.getActivityControlSurface().attachToActivity(
          host.getActivity(),
          host.getLifecycle()
      );
 and
 host.configureFlutterEngine(flutterEngine); 
will call FlutterEnginePluginRegistry#attachToActivity
 activityAware.onAttachedToActivity(activityPluginBinding);

However, when I do a code search for `attachToActivity()` I only find that method invoked from one place, which is within `onAttach()`:

https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java#L178

And `onAttach()` is only invoked from independent places:

`FlutterActivity`
https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterActivity.java#L418

`FlutterFragment`
https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterFragment.java#L596

Can you clarify where you think the problem is, or provide a minimal GitHub project that reproduces your issue?

**Comment by matthew-carroll** [comment]
@mikemaksymowych can you check to see if your plugin is being registered twice? We're aware of an issue with repeated plugin registrations which is being fixed here: 

https://github.com/flutter/flutter/issues/49365

**Comment by mikemaksymowych** [comment]
@matthew-carroll thanks, that sounds more like my issue.

**Comment by v7lin** [comment]
why not create a new plugin、add activityware，and debug

**Comment by matthew-carroll** [comment]
@v7lin I wasn't able to reproduce your result, that's why I asked about the code path. Can you confirm whether or not your problem is the same as @mikemaksymowych which is actually a double registration problem as described in #49365 ?

**Comment by v7lin** [comment]
Demo
` ` `shell
git clone -b v2 https://github.com/v7lin/fake_tencent.git
` ` `

**Comment by matthew-carroll** [comment]
@v7lin I cloned your repo and ran the app. Unfortunately, I can't read the text in the app and I'm not sure how to reproduce the issue.

However, I created a custom plugin project and traced `Activity` attachment and the  `onActivityResult()`. Based on what I'm seeing, the issue that you're experiencing really is the same as #49365. When I run against the latest version of the Flutter engine, the duplicate `onActivityResult()` call goes away.

I'm going to tentatively close this issue as resolved in #49365. If you find any direct evidence that this problem still exists after the Flutter framework receives the fix in #49365 please feel free to re-open this issue with a reproduction of that problem. Thanks.

**Comment by shingohu** [comment]
@v7lin Version 1.12 will register the plug-in automatically by the reflection. Two registration problems. Because in the host Activity using GeneratedPluginRegistrant registerWith (flutterEngine); Comment out this line of code
